keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,354 @@
+type KubeletConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
+    ------------
+
+    // +optional
+    // TracingConfiguration
+    TracingConfiguration componentbaseconfig.TracingConfiguration
+}
+
+// TracingConfiguration specifies configuration for opentelemetry tracing
+type TracingConfiguration struct {
+    // +optional
+    // URL of the collector that's running on the node.
+    // Defaults to 0.0.0.0:4317
+    URL  *string `json:"url,omitempty" protobuf:"bytes,1,opt,name=url"`
+
+    // +optional
+    // SamplingRatePerMillion is the number of samples to collect per million spans.
+    // Defaults to 0.
+    SamplingRatePerMillion int32 `json:"samplingRatePerMillion,omitempty" protobuf:"varint,2,opt,name=samplingRatePerMillion"`
+}
+
+```
+If a TracingConfiguration is not specified, the kubelet will not send any spans, even if incoming requests ask for sampling.
+
+## Design Details
+
+### Test Plan
+
+We will test tracing added by this feature with an integration test.  The
+integration test will verify that spans exported by the kubelet match what is
+expected from the request. We will also add an integration test that verifies
+spans propagated from kubelet to API server match what is expected from the request.
+
+### Graduation Requirements
+
+Alpha
+
+- [] Implement tracing of incoming and outgoing gRPC, HTTP requests in the kubelet
+- [] Integration testing of tracing
+
+Beta
+
+- [] Publish examples of how to use the OT Collector with kubernetes
+- [] Allow time for feedback
+- [] Revisit the format used to export spans.
+
+GA
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+  - [X] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: KubeletTracing
+    - Components depending on the feature gate: kubelet
+  - [X] Other
+    - Describe the mechanism: Kubelet Configuration.
+    - Will enabling / disabling the feature require downtime of the control
+      plane?  **No. It will require restarting the kubelet service per node.**
+    - Will enabling / disabling the feature require downtime or reprovisioning
+      of a node? **No, restarting the kubelet with feature-gate disabled will disable tracing**
+
+##### Does enabling the feature change any default behavior?
+  No. The feature is disabled unlesss the feature gate is enabled and the TracingConfiguration is populated in Kubelet Configuration.
+  When the feature is enabled, it doesn't change behavior from the users' perspective; it only adds tracing telemetry.
+
+##### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+  Yes.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+  It will start generating and exporting traces again.
+
+##### Are there any tests for feature enablement/disablement?
+  Unit tests switching feature gates will be added.
```

Comment:
You may also want to perform manual testing where you disable and reenable the feature on a node and ensure the kubelet still comes up.

Reply:
added